Controllers/HierarchyTerritoriesAdminController.cs

```diff
@@ -0,0 +1,525 @@
+                        : T("Your {0} has been created.", item.TypeDefinition.DisplayName));
+
+                    return this.RedirectLocal(returnUrl, () => 
+                        RedirectToAction("EditTerritory", 
+                            new RouteValueDictionary { { "Id", item.Id } }));
+                }
+            });
+        }
+
+        #endregion
+
+        #region Edit
+        [HttpGet]
+        public ActionResult EditTerritory(int id) {
+
+            var territoryItem = _contentManager.Get(id, VersionOptions.Latest);
+            if (territoryItem == null)
+                return HttpNotFound();
+            var territoryPart = territoryItem.As<TerritoryPart>();
+            if (territoryPart == null)
+                return HttpNotFound();
+
+            ActionResult redirectTo;
+            if (ShouldRedirectForPermissions(territoryPart.Record.Hierarchy.Id, out redirectTo)) {
+                return redirectTo;
+            }
+            
+            if (!_authorizer.Authorize(
+                Orchard.Core.Contents.Permissions.EditContent, territoryItem, TerritoriesUtilities.Edit401TerritoryMessage))
+                return new HttpUnauthorizedResult();
+
+            // We should have filtered out the cases where we cannot or should not be editing the item here
+            var model = _contentManager.BuildEditor(territoryItem);
+            return View(model.Hierarchy(territoryItem.As<TerritoryPart>().Hierarchy));
+        }
+
+        [HttpPost, ActionName("EditTerritory")]
+        [Orchard.Mvc.FormValueRequired("submit.Save")]
+        public ActionResult EditTerritoryPost(int id, string returnUrl) {
+            return EditTerritoryPost(id, returnUrl, contentItem => {
+                if (!contentItem.Has<IPublishingControlAspect>() && 
+                    !contentItem.TypeDefinition.Settings.GetModel<ContentTypeSettings>().Draftable)
+                    _contentManager.Publish(contentItem);
+            });
+        }
+
+        [HttpPost, ActionName("EditTerritory")]
+        [Orchard.Mvc.FormValueRequired("submit.Publish")]
+        public ActionResult EditAndPublishTerritoryPost(int id, string returnUrl) {
+            var content = _contentManager.Get(id, VersionOptions.Latest);
+
+            if (content == null)
+                return HttpNotFound();
+
+            if (!_authorizer.Authorize(
+                Orchard.Core.Contents.Permissions.PublishContent, content, TerritoriesUtilities.Edit401TerritoryMessage))
+                return new HttpUnauthorizedResult();
+
+            return EditTerritoryPost(id, returnUrl, contentItem => _contentManager.Publish(contentItem));
+        }
+
+        private ActionResult EditTerritoryPost(
+            int id, string returnUrl, Action<ContentItem> conditionallyPublish) {
+
+            var territoryItem = _contentManager.Get(id, VersionOptions.DraftRequired);
+            return ExecuteTerritoryPost(new TerritoryExecutionContext {
```

Comment:
tldr: I'll try.
Full answer:
This big return statement (along with similar ones in the same controller) is an attempt to refactor stuff already. Several checks and stuff are common among the post methods. However, and naturally, Edit, Create and Delete have their own stuff to do, and that is being passed as a Func<ContentItem, ActionResult> in the context.
The ExecuteTerritoryPost method takes the context, does all the prechecks (that are common among the post methods), then executes the specific functionalities.

Another way to achieve the same would have looked similar to the ShouldRedirectForPermissions method from the same controller, that is basically designed on the same logic as doing `if(TrySomething(parameter, out eventualResult) { }`.

Since in the case of the post methods now calling ExecuteTerritoriesPost that call happens at the end, changing to the other "style" wouldn't make the method's body much shorter.

Anyway, I removed from the Func in the Edit the code where I was fiddling with the returnUrl. Upon rereading it, I was wondering why I wrote it in the first place, and could not figure it out, since it was not doing anything usefult anyway, but rather trying to change the returnUrl to the url for the display of the ContentItem.